modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy

```diff
@@ -342,4 +363,52 @@ abstract class RepositoryProvider {
         }
     }
 
+    /**
+     * Creates a retry policy using the configuration specified by {@link nextflow.scm.ProviderRetryConfig}
+     *
+     * @param cond A predicate that determines when a retry should be triggered
+     * @return The {@link dev.failsafe.RetryPolicy} instance
+     */
+    protected <T> RetryPolicy<T> retryPolicy(Predicate<? extends Throwable> cond) {
+        final listener = new EventListener<ExecutionAttemptedEvent<?>>() {
+            @Override
+            void accept(ExecutionAttemptedEvent<?> event) throws Throwable {
+                log.debug("Provider invoke response error - attempt: ${event.attemptCount}; reason: ${event.lastFailure.message}")
+            }
+        }
+        return RetryPolicy.<T>builder()
+            .handleIf(cond)
+            .withBackoff(retryConfig.delay.toMillis(), retryConfig.maxDelay.toMillis(), ChronoUnit.MILLIS)
+            .withMaxAttempts(retryConfig.maxAttempts)
+            .withJitter(retryConfig.jitter)
+            .onRetry(listener as EventListener)
+            .build()
+    }
+
+    /**
+     * Carry out the invocation of the specified action using a retry policy.
+     *
+     * @param action A {@link dev.failsafe.function.CheckedSupplier} instance modeling the action to be performed in a safe manner
+     * @return The result of the supplied action
+     */
+    protected <T> T apply(CheckedSupplier<T> action) {
+        // define the retry condition
+        final cond = new Predicate<? extends Throwable>() {
+            @Override
+            boolean test(Throwable t) {
+                // checkResponse method can throw RateLimitExceedExceptions which are retried and AbortedException which not retried.
+                // Other methods can throw IOExceptions including for 50x response codes which are retried by default.
+                if( t instanceof RateLimitExceededException || t.cause instanceof RateLimitExceededException)
+                    return true
+                if( t instanceof IOException || t.cause instanceof IOException )
```

Comment:
`IOException` is too generic. It would be preferable having a more granular control on http error codes [here](https://github.com/nextflow-io/nextflow/blob/4b4568065ef78af080d4f3e2116a75fdb381d0ab/modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy#L253-L253). 

However it can be difficult to achieve this in the current implementation. Wonder if it should be used a different http client (see WaveClient)